server.go

```diff
@@ -55,6 +62,98 @@ type patchOperation struct {
 	Value interface{} `json:"value,omitempty"`
 }
 
+// watchNamespaces monitors newly created namespaces. On namespace creation, an image pull secret
+// will be created in the namespace to access GCR and AR registries. Any previously created
+// namespaces will also get the secret.
+func watchNamespaces() error {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return fmt.Errorf("getting cluster config: %v", err)
+	}
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("getting clientset: %v", err)
+	}
+
+	// grab credentials from where GCP would normally look
+	ctx := context.Background()
+	creds, err := google.FindDefaultCredentials(ctx)
+	if err != nil {
+		return fmt.Errorf("finding default credentials: %v", err)
+	}
+
+	watcher, err := clientset.CoreV1().Namespaces().Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("creating namespace watcher: %v", err)
+	}
+	for e := range watcher.ResultChan() {
+		ns := e.Object.(*corev1.Namespace)
+		if e.Type == watch.Added {
+			if err := createPullSecret(clientset, ns, creds); err != nil {
+				log.Printf("creating pull secret: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// createPullSecret creates an image registry pull secret to the provided namespace using the provided creds.
+func createPullSecret(clientset *kubernetes.Clientset, ns *corev1.Namespace, creds *google.Credentials) error {
+	if skipNamespace(ns.Name) {
+		return nil
+	}
+
+	token, err := creds.TokenSource.Token()
+	if err != nil {
+		return err
+	}
+
+	var dockercfg string
+	registries := append(gcr_config.DefaultGCRRegistries[:], gcr_config.DefaultARRegistries[:]...)
+	for _, reg := range registries {
+		dockercfg += fmt.Sprintf(`"https://%s":{"username":"oauth2accesstoken","password":"%s","email":"none"},`, reg, token.AccessToken)
+	}
+	dockercfg = strings.TrimSuffix(dockercfg, ",")
+	data := map[string][]byte{
+		".dockercfg": []byte(fmt.Sprintf(`{%s}`, dockercfg)),
+	}
+
+	secrets := clientset.CoreV1().Secrets(ns.Name)
+
+	exists := false
+	secList, err := secrets.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, s := range secList.Items {
+		if s.Name == gcpAuth {
+			exists = true
+			break
+		}
+	}
+
+	if !exists {
```

Comment:
Could inverse this `if` and just `return nil` if `exists` to prevent the large nested block, but just a suggestion